rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+  ```
+  PrimitivePartitionRule
+    OnlyValidPartitionRule
+      CombinePartitionRule (with a join-anything combiner rule)
+        OpPredicatePartitionRule
+  ```
+
+- Classic pattern-based BYOC with `MergeComposite`/`AnnotateTarget`/`PartitionGraph` passes
+  (eg see `cutlass.py`)`:
+  ```
+  PrimitivePartitionRule
+    OnlyValidPartitionRule
+      CombinePartitionRule (with join-anything combiner rule)
+        UnionPartitionRule
+          CompositePartitionRule(label1)
+            DFPatternPartitionRule(pattern1)
+                      :
+          CompositePartitionRule(labeln)
+            DFPatternPartitionRule(patternn)
+  ```
+
+  The `CompositePartitionRule`/`DFPatternPartitionRule` combination is repeated for each entry in the pattern table for
+  the BYOC toolchain name, eg:
+  ```
+  CompositePartitionRule(
+    rule_name="cutlass.conv2d_bias_residual_multiply_relu"
+    sub_rule=DFPatternPartitionRule(
+      pattern=CallPatternNode(Op(nn.relu), 
+                              [AltPattern(CallPatternNode(Op(multiply), 
+                                                          [CallPatternNode(AltPattern(Op(add) | Op(nn.bias_add)),
+                                                                           [CallPatternNode(Op(nn.conv2d), [*, *]), *]),
+                                                           *]) |
+                                          CallPatternNode(Op(multiply),
+                                                          [*, 
+                                                           CallPatternNode(AltPattern(Op(add) | Op(nn.bias_add)),
+                                                                           [CallPatternNode(Op(nn.conv2d), [*, *]), *])
+                                                          ]))
+                              ])))
+  ```
+
+- "Consider this library implementation for these sub-expressions", using `DFPatterns` to pick out which Relay operators
+  are supported (a new scheme):
+  ```
+  OnlyValidPartitionRule
+    CombinePartitionRule (with default TVM combiner rules)
+      UnionPartitionRule
+        OpCallByKindPartitionRule
+        CompositePartitionRule(lable1)
+          DFPatternPartitionRule(pattern1)
+                     :
+        CompositePartitionRule(lablen)
+          DFPatternPartitionRule(patternn)
+  ```
+
+- Classic TVM `FuseOps`:
+  ```
+  PrimitivePartitionRule
+    OnlyValidPartitionRule
+      CombinePartitionRule (with default TVM combiner rules)
+        OpCallByKindPartitionRule
+  ```
+
+- "Just fuse what I tell you to fuse", using `DFPatterns` to directly select candidates (a new scheme):
+  ```
+  PrimitivePartitionRule
+    OnlyValidPartitionRule
+      UnionPartitionRule
+        DFPatternPartitionRule(pattern1)
+                     :
+        DFPatternPartitionRule(patternn)
+  ```
+
+### PartitionSpec
+
+A `PartitionSpec` pairs a a `PartitionRule` with one or more `Target`s.
```

Comment:
I suppose this is related to the above question.

Reply:
It's really just here for book-keeping, all the real action takes place with PartitionRules.